destination_test.go

```
@@ -15,92 +30,150 @@ import (
+	// Pinecone writes appear to be asynchronous. At the very least, in the current free tier serverless
+	// configuration that I've tested, pinecone writes occurred slightly after the RPC call
+	// returned data. Therefore, the following retry logic is needed to make tests more robust
+	for i := 1; i <= maxRetries; i++ {
+		res, err := index.FetchVectors(&ctx, []string{id})
+		is.NoErr(err)
+
+		vec, ok := res.Vectors[id]
+		if !ok {
+			if i == maxRetries {
+				is.Fail() // vector was not written
+			} else {
+				wait := waitTime(i)
+				t.Logf("retrying with wait of %v", wait)
+				time.Sleep(wait)
+				continue
+			}
+		}
+
+		is.Equal(vec.Values, writtenVecs.Values)
+		is.Equal(vec.SparseValues.Values, writtenVecs.SparseValues.Values)
+		is.Equal(vec.SparseValues.Indices, writtenVecs.SparseValues.Indices)
+		break
 	}
+}
 
-	recVecValues, err := parseRecordPayload(rec.Payload)
-	is.NoErr(err)
+func assertDeletedRecordIndex(ctx context.Context, t *testing.T, is *is.I, index *pinecone.IndexConnection, id string) {
+	// same as assertWrittenRecordIndex, we need the retry for robustness
+	for i := 0; i <= maxRetries; i++ {
+		res, err := index.FetchVectors(&ctx, []string{id})
+		is.NoErr(err)
+
+		_, ok := res.Vectors[id]
+		if ok {
+			if i == maxRetries {
+				is.Fail() // vector found, not properly deleted
+			} else {
+				wait := waitTime(i)
+				t.Logf("retrying with wait of %v", wait)
+				time.Sleep(wait)
+				continue
+			}
+		}
+	}
+}
 
-	is.Equal(vec.Values, recVecValues.Values)
-	is.Equal(vec.SparseValues.Indices, recVecValues.SparseValues.Indices)
-	is.Equal(vec.SparseValues.Values, recVecValues.SparseValues.Values)
+func deleteAllRecords(is *is.I, index *pinecone.IndexConnection) {
+	ctx := context.Background()
+	err := index.DeleteAllVectorsInNamespace(&ctx)
+	is.NoErr(err)
 }
 
 func TestMain(t *testing.M) {
-	if err := godotenv.Load(); err != nil {
-		log.Fatal("error loading environment variables")
-	}
+	godotenv.Load()
```

I didn't use this library before, it looks like it's about loading `.env` files. However, I can't see any of those in this repository. Is it about something else?

---

Good reminder that I have to add this into the docs. You can see that the `.env` file is gitignored. When developing locally, one would have to obtain the pinecone host and api key and put them inside the .env file as env variables.
`godotenv` loads them up into the local environment. On CI there's no env file, so the error is ignored.